go/api/config.go

```diff
@@ -0,0 +1,290 @@
+}
+
+// RedisCredentials represents the credentials for connecting to a Redis server.
+type RedisCredentials struct {
+	// Optional: the username that will be used for authenticating connections to the Redis servers. If not supplied, "default"
+	// will be used.
+	username string
+	// Required: the password that will be used for authenticating connections to the Redis servers.
+	password string
+}
+
+// NewRedisCredentials returns a new RedisCredentials struct with the given username and password.
+func NewRedisCredentials(username string, password string) *RedisCredentials {
+	return &RedisCredentials{username, password}
+}
+
+// NewRedisCredentialsWithDefaultUsername returns a new RedisCredentials struct with a default username of "default" and the
+// given password.
+func NewRedisCredentialsWithDefaultUsername(password string) *RedisCredentials {
+	return &RedisCredentials{password: password}
+}
+
+func (creds *RedisCredentials) toProtobuf() *protobuf.AuthenticationInfo {
+	return &protobuf.AuthenticationInfo{Username: creds.username, Password: creds.password}
+}
+
+// ReadFrom represents the client's read from strategy.
+type ReadFrom int
+
+const (
+	// Primary - Always get from primary, in order to get the freshest data.
+	Primary ReadFrom = 0
+	// PreferReplica - Spread the requests between all replicas in a round-robin manner. If no replica is available, route the
+	// requests to the primary.
+	PreferReplica ReadFrom = 1
+)
+
+func mapReadFrom(readFrom ReadFrom) protobuf.ReadFrom {
+	if readFrom == PreferReplica {
+		return protobuf.ReadFrom_PreferReplica
+	}
+
+	return protobuf.ReadFrom_Primary
+}
+
+type baseClientConfiguration struct {
+	addresses      []NodeAddress
+	useTLS         bool
+	credentials    *RedisCredentials
+	readFrom       ReadFrom
+	requestTimeout int
+}
+
+func (config *baseClientConfiguration) toProtobuf() *protobuf.ConnectionRequest {
+	request := protobuf.ConnectionRequest{}
+	for _, address := range config.addresses {
+		request.Addresses = append(request.Addresses, address.toProtobuf())
+	}
+
+	if config.useTLS {
+		request.TlsMode = protobuf.TlsMode_SecureTls
+	} else {
+		request.TlsMode = protobuf.TlsMode_NoTls
+	}
+
+	if config.credentials != nil {
+		request.AuthenticationInfo = config.credentials.toProtobuf()
+	}
+
+	request.ReadFrom = mapReadFrom(config.readFrom)
+	if config.requestTimeout != 0 {
+		request.RequestTimeout = uint32(config.requestTimeout)
+	}
+
+	return &request
+}
+
+// BackoffStrategy represents the strategy used to determine how and when to reconnect, in case of connection failures. The
+// time between attempts grows exponentially, to the formula:
+//
+//	rand(0 ... factor * (exponentBase ^ N))
+//
+// where N is the number of failed attempts.
+//
+// Once the maximum value is reached, that will remain the time between retry attempts until a reconnect attempt is successful.
+// The client will attempt to reconnect indefinitely.
+type BackoffStrategy struct {
```

Comment:
@aaron-congo can you confirm that the formatting for these comments is preserved when generating documentation?

Reply:
The line length is not preserved, but blank lines and indents are. So docs for the above look like this (you can get them by running `go doc -all github.com/aws/glide-for-redis/go/glide/api`
```
    BackoffStrategy represents the strategy used to determine how and when to
    reconnect, in case of connection failures. The time between attempts grows
    exponentially, to the formula:

        rand(0 ... factor * (exponentBase ^ N))

    where N is the number of failed attempts.

    Once the maximum value is reached, that will remain the time between retry
    attempts until a reconnect attempt is successful. The client will attempt to
    reconnect indefinitely.
```

Reply:
With Go docs you create code blocks by indenting the relevant code, like I've done here. There is no inline code highlighting like you would get with backticks. When you upload the package the site will automatically render the Go docs into html, which includes rendering the indented code blocks as <pre> blocks